File: lib/android/vesper-player-kit/src/main/java/io/github/ikaros/vesper/player/android/VesperPlayerError.kt
package io.github.ikaros.vesper.player.android

enum class VesperPlayerErrorCode(
    val wireName: String,
    internal val jniOrdinal: Int,
) {
    InvalidArgument("invalidArgument", 0),
    InvalidState("invalidState", 1),
    InvalidSource("invalidSource", 2),
    BackendFailure("backendFailure", 3),
    AudioOutputUnavailable("audioOutputUnavailable", 4),
    DecodeFailure("decodeFailure", 5),
    SeekFailure("seekFailure", 6),
    Unsupported("unsupported", 7),
    CommandChannelClosed("commandChannelClosed", 8),
    EventChannelClosed("eventChannelClosed", 9),
    Cancelled("cancelled", 10),
    Timeout("timeout", 11);

    companion object {
        fun fromWireName(wireName: String?): VesperPlayerErrorCode =
            entries.firstOrNull { it.wireName == wireName } ?: BackendFailure

        internal fun fromJniOrdinal(ordinal: Int): VesperPlayerErrorCode =
            entries.firstOrNull { it.jniOrdinal == ordinal } ?: BackendFailure
    }
}

enum class VesperPlayerErrorCategory(
    val wireName: String,
    internal val jniOrdinal: Int,
) {
    Input("input", 0),
    Source("source", 1),
    Network("network", 2),
    Decode("decode", 3),
    AudioOutput("audioOutput", 4),
    Playback("playback", 5),
    Capability("capability", 6),
    Platform("platform", 7);

    companion object {
        fun fromWireName(wireName: String?): VesperPlayerErrorCategory =
            entries.firstOrNull { it.wireName == wireName } ?: Platform

        internal fun fromJniOrdinal(ordinal: Int): VesperPlayerErrorCategory =
            entries.firstOrNull { it.jniOrdinal == ordinal } ?: Platform
    }
}

data class VesperPlayerErrorState(
    val message: String,
    val code: VesperPlayerErrorCode,
    val category: VesperPlayerErrorCategory,
    val retriable: Boolean,
    val details: Map<String, Any?> = emptyMap(),
) {
    fun toMap(): Map<String, Any?> =
        mapOf(
            "message" to message,
            "code" to code.wireName,
            "category" to category.wireName,
            "retriable" to retriable,
            "details" to details,
        )
}

File: lib/ios/VesperPlayerKit/Sources/VesperPlayerKit/Bridge/VesperPlayerError.swift
import AVFoundation
import CoreGraphics
import Foundation
import SwiftUI
import UIKit
internal import VesperPlayerKitBridgeShim
public enum VesperPlayerErrorCode: String, Equatable, Codable {
    case invalidArgument
    case invalidState
    case invalidSource
    case backendFailure
    case audioOutputUnavailable
    case decodeFailure
    case seekFailure
    case unsupported
    case commandChannelClosed
    case eventChannelClosed
    case cancelled
    case timeout
}

public enum VesperPlayerErrorCategory: String, Equatable, Codable {
    case input
    case source
    case network
    case decode
    case audioOutput
    case playback
    case capability
    case platform
}

extension VesperPlayerErrorCode {
    init(ffiCode: PlayerFfiErrorCode) {
        switch ffiCode {
        case PlayerFfiErrorCodeInvalidArgument, PlayerFfiErrorCodeNullPointer,
             PlayerFfiErrorCodeInvalidUtf8, PlayerFfiErrorCodeNone:
            self = .invalidArgument
        case PlayerFfiErrorCodeInvalidState:
            self = .invalidState
        case PlayerFfiErrorCodeInvalidSource:
            self = .invalidSource
        case PlayerFfiErrorCodeBackendFailure:
            self = .backendFailure
        case PlayerFfiErrorCodeAudioOutputUnavailable:
            self = .audioOutputUnavailable
        case PlayerFfiErrorCodeDecodeFailure:
            self = .decodeFailure
        case PlayerFfiErrorCodeSeekFailure:
            self = .seekFailure
        case PlayerFfiErrorCodeUnsupported:
            self = .unsupported
        case PlayerFfiErrorCodeCommandChannelClosed:
            self = .commandChannelClosed
        case PlayerFfiErrorCodeEventChannelClosed:
            self = .eventChannelClosed
        case PlayerFfiErrorCodeCancelled:
            self = .cancelled
        case PlayerFfiErrorCodeTimeout:
            self = .timeout
        default:
            self = .backendFailure
        }
    }

    var ffiCode: PlayerFfiErrorCode {
        switch self {
        case .invalidArgument: return PlayerFfiErrorCodeInvalidArgument
        case .invalidState: return PlayerFfiErrorCodeInvalidState
        case .invalidSource: return PlayerFfiErrorCodeInvalidSource
        case .backendFailure: return PlayerFfiErrorCodeBackendFailure
        case .audioOutputUnavailable: return PlayerFfiErrorCodeAudioOutputUnavailable
        case .decodeFailure: return PlayerFfiErrorCodeDecodeFailure
        case .seekFailure: return PlayerFfiErrorCodeSeekFailure
        case .unsupported: return PlayerFfiErrorCodeUnsupported
        case .commandChannelClosed: return PlayerFfiErrorCodeCommandChannelClosed
        case .eventChannelClosed: return PlayerFfiErrorCodeEventChannelClosed
        case .cancelled: return PlayerFfiErrorCodeCancelled
        case .timeout: return PlayerFfiErrorCodeTimeout
        }
    }
}

extension VesperPlayerErrorCategory {
    init(ffiCategory: PlayerFfiErrorCategory) {
        switch ffiCategory {
        case PlayerFfiErrorCategoryInput:
            self = .input
        case PlayerFfiErrorCategorySource:
            self = .source
        case PlayerFfiErrorCategoryNetwork:
            self = .network
        case PlayerFfiErrorCategoryDecode:
            self = .decode
        case PlayerFfiErrorCategoryAudioOutput:
            self = .audioOutput
        case PlayerFfiErrorCategoryPlayback:
            self = .playback
        case PlayerFfiErrorCategoryCapability:
            self = .capability
        case PlayerFfiErrorCategoryPlatform:
            self = .platform
        default:
            self = .platform
        }
    }

    var ffiCategory: PlayerFfiErrorCategory {
        switch self {
        case .input: return PlayerFfiErrorCategoryInput
        case .source: return PlayerFfiErrorCategorySource
        case .network: return PlayerFfiErrorCategoryNetwork
        case .decode: return PlayerFfiErrorCategoryDecode
        case .audioOutput: return PlayerFfiErrorCategoryAudioOutput
        case .playback: return PlayerFfiErrorCategoryPlayback
        case .capability: return PlayerFfiErrorCategoryCapability
        case .platform: return PlayerFfiErrorCategoryPlatform
        }
    }
}

public struct VesperPlayerError: Error, Equatable {
    public let message: String
    public let code: VesperPlayerErrorCode
    public let category: VesperPlayerErrorCategory
    public let retriable: Bool
    public let details: [String: String]

    public init(
        message: String,
        code: VesperPlayerErrorCode,
        category: VesperPlayerErrorCategory,
        retriable: Bool,
        details: [String: String] = [:]
    ) {
        self.message = message
        self.code = code
        self.category = category
        self.retriable = retriable
        self.details = details
    }
}
